File: regions/system/library/go/building-blocks/registry_test.go
package buildingblocks

import (
	"context"
	"testing"
)

// テスト用のコンポーネント実装。
type testComponent struct {
	name    string
	version string
}

func newTestComponent(name string) *testComponent {
	return &testComponent{name: name, version: "1.0.0"}
}

func (c *testComponent) Name() string { return c.name }
func (c *testComponent) Version() string { return c.version }
func (c *testComponent) Init(_ context.Context, _ Metadata) error { return nil }
func (c *testComponent) Close(_ context.Context) error { return nil }
func (c *testComponent) Status(_ context.Context) ComponentStatus { return StatusReady }

// コンポーネントを登録して名前で取得できることを確認する。
func TestRegistry_RegisterAndGet(t *testing.T) {
	r := NewComponentRegistry()
	c := newTestComponent("comp-1")

	if err := r.Register(c); err != nil {
		t.Fatalf("Register failed: %v", err)
	}

	got, ok := r.Get("comp-1")
	if !ok {
		t.Fatal("Get returned false for registered component")
	}
	if got.Name() != "comp-1" {
		t.Errorf("expected name 'comp-1', got '%s'", got.Name())
	}
}

// 同名のコンポーネントを重複登録するとエラーが返ることを確認する。
func TestRegistry_RegisterDuplicate(t *testing.T) {
	r := NewComponentRegistry()
	r.Register(newTestComponent("dup"))

	err := r.Register(newTestComponent("dup"))
	if err == nil {
		t.Fatal("expected error on duplicate registration, got nil")
	}
}

// 存在しない名前でコンポーネントを取得すると false が返ることを確認する。
func TestRegistry_GetNotFound(t *testing.T) {
	r := NewComponentRegistry()
	_, ok := r.Get("missing")
	if ok {
		t.Fatal("expected false for missing component")
	}
}

// InitAll が登録済みの全コンポーネントを初期化することを確認する。
func TestRegistry_InitAll(t *testing.T) {
	r := NewComponentRegistry()
	r.Register(newTestComponent("a"))
	r.Register(newTestComponent("b"))

	if err := r.InitAll(context.Background()); err != nil {
		t.Fatalf("InitAll failed: %v", err)
	}
}

// CloseAll が登録済みの全コンポーネントをクローズすることを確認する。
func TestRegistry_CloseAll(t *testing.T) {
	r := NewComponentRegistry()
	r.Register(newTestComponent("a"))

	if err := r.CloseAll(context.Background()); err != nil {
		t.Fatalf("CloseAll failed: %v", err)
	}
}

// StatusAll が全コンポーネントのステータスを返すことを確認する。
func TestRegistry_StatusAll(t *testing.T) {
	r := NewComponentRegistry()
	r.Register(newTestComponent("a"))
	r.Register(newTestComponent("b"))

	statuses := r.StatusAll(context.Background())
	if len(statuses) != 2 {
		t.Fatalf("expected 2 statuses, got %d", len(statuses))
	}
	for _, name := range []string{"a", "b"} {
		if statuses[name] != StatusReady {
			t.Errorf("expected Ready for '%s', got '%s'", name, statuses[name])
		}
	}
}

File: regions/system/library/typescript/building-blocks/src/registry.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { ComponentRegistry } from './registry.js';
import type { Component, ComponentStatus } from './component.js';

// テスト用のシンプルなコンポーネント実装。
class TestComponent implements Component {
  readonly name: string;
  readonly componentType = 'test';
  private _status: ComponentStatus = 'uninitialized';

  constructor(name: string) {
    this.name = name;
  }

  async init(): Promise<void> {
    this._status = 'ready';
  }

  async close(): Promise<void> {
    this._status = 'closed';
  }

  async status(): Promise<ComponentStatus> {
    return this._status;
  }

  metadata(): Record<string, string> {
    return {};
  }
}

// ComponentRegistry のテスト: 登録・取得・重複エラー・一括操作の動作を検証する。
describe('ComponentRegistry', () => {
  let registry: ComponentRegistry;

  beforeEach(() => {
    registry = new ComponentRegistry();
  });

  // コンポーネントを登録して名前で取得できることを確認する。
  it('コンポーネントを登録して取得できること', () => {
    const c = new TestComponent('comp-1');
    registry.register(c);

    const got = registry.get('comp-1');
    expect(got).toBe(c);
  });

  // 同名のコンポーネントを重複登録するとエラーがスローされることを確認する。
  it('同名コンポーネントの重複登録でエラーがスローされること', () => {
    registry.register(new TestComponent('dup'));
    expect(() => registry.register(new TestComponent('dup'))).toThrow(
      "コンポーネント 'dup' は既に登録されています"
    );
  });

  // 存在しない名前でコンポーネントを取得すると undefined が返ることを確認する。
  it('未登録のコンポーネントは undefined を返すこと', () => {
    expect(registry.get('missing')).toBeUndefined();
  });

  // initAll が全コンポーネントを初期化することを確認する。
  it('initAll で全コンポーネントが初期化されること', async () => {
    const a = new TestComponent('a');
    const b = new TestComponent('b');
    registry.register(a);
    registry.register(b);

    await registry.initAll();

    expect(await a.status()).toBe('ready');
    expect(await b.status()).toBe('ready');
  });

  // closeAll が全コンポーネントをクローズすることを確認する。
  it('closeAll で全コンポーネントがクローズされること', async () => {
    const a = new TestComponent('a');
    registry.register(a);
    await a.init();

    await registry.closeAll();

    expect(await a.status()).toBe('closed');
  });

  // statusAll が全コンポーネントのステータスを返すことを確認する。
  it('statusAll が全コンポーネントのステータスを返すこと', async () => {
    registry.register(new TestComponent('a'));
    registry.register(new TestComponent('b'));
    await registry.initAll();

    const statuses = await registry.statusAll();
    expect(statuses['a']).toBe('ready');
    expect(statuses['b']).toBe('ready');
  });
});
